### ISSUE TITLE
Towards an H5Easy opencv support (incl. rough prototype for H5Easy::load)

### ISSUE THREAD

**Issue by spacescientist** [BlueBrain/HighFive#294]
Incidentally, since I was so happy about #286 and since we also mentioned a possible opencv overload, I've given a try and somehow managed to make HighFive read a dataset from an HDF file and store it as an opencv `cv::Mat_<double>`.

It's not nice at all as it stands: it would still need to be generalised beyond `double` and it's moreover incomplete (since there is only `load` for now; I haven't touched `dump`), but it seems to work in the simple setting I used. I'm surprised `cv::Mat_` understood `rows` and `cols`, but it apparently did.

##### H5Easy_opencv.hpp (prototype)
```cpp
#ifndef H5EASY_BITS_OPENCV_HPP
#define H5EASY_BITS_OPENCV_HPP

#include "../H5Easy.hpp"
#include "H5Easy_misc.hpp"
#include "H5Easy_scalar.hpp"  // to get the basic "load_impl"

////#ifdef H5_USE_OPENCV

namespace H5Easy {

namespace detail {

namespace opencv {

// return the shape of the opencv-object as "std::vector<size_t>"
template <class T>
inline std::vector<size_t> shape(const T& data)
{
    return std::vector<size_t>(data.shape().cbegin(), data.shape().cend());
}


// load opencv-object from DataSet
template <class T>
struct load_impl
{
    static T run(const File& file, const std::string& path)
    {
        DataSet dataset = file.getDataSet(path);
        std::vector<size_t> dims = dataset.getDimensions();
        T data(dims[0], dims[1]);
        dataset.read(reinterpret_cast<double*>(data.data));
        //                             ^ ! Shouldn't be explicitly double of course TODO.
        return data;
    }
};

}  // namespace opencv

// front-end
template <class T>
struct load_impl<cv::Mat_<T>>
{
    static cv::Mat_<T> run(const File& file, const std::string& path)
    {
        return detail::opencv::load_impl<cv::Mat_<T>>::run(file, path);
    }
};


}  // namespace detail

}  // namespace H5Easy

////////#endif  // H5_USE_OPENCV

#endif  // H5EASY_BITS_OPENCV_HPP
```

##### H5Easy.hpp
```cpp
// I simply added at the end of H5Easy.hpp:
#include "h5easy_bits/H5Easy_opencv.hpp"
```

And modified the MWE to:
##### main.cpp
```cpp
#include <iostream>
#include <Eigen/Eigen>
#include <opencv2/core/eigen.hpp>
#include "highfive/H5Easy.hpp"

int main( int argc, char** argv )
{

	H5Easy::File input_hdf("out_test.hdf", H5Easy::File::ReadOnly);

	// Works!
	auto data_cv = H5Easy::load<cv::Mat_<double>>(input_hdf, "/EigenRowMajorFromCV");

	std::cout << data_cv << std::endl;
	std::cout << data_cv.rows << std::endl;
	std::cout << data_cv.row(2) << std::endl;

	return 0;
}

```

Could it be interesting for you to expend on this?

Cheers.

**Comment by tdegeus** [comment]
Nice, Why don't you open a pull request such that we can take it from there.

**Comment by spacescientist** [comment]
Alright :) 

Again, beware that while it behaves as expected with cv::Mat<double>, this is nothing more than a first attempt/dirty hack, which adamantly requires fixing. It is still so rough that `double` is actually still hardcoded in the prototype H5Easy_opencv.hpp.

**Comment by tdegeus** [comment]
No worries that it still needs to be polished. We'll get there with some reviewing

**Comment by eudoxos** [comment]
This can be closed, is superceded by #295.
